### PR TITLE
Add autofill subfeature feature-flagging to remote config

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/488551667048375/1203850982819858/f

### Description
Consumes privacy remote config to obtain autofill _subfeature_ configuration. This allows `autofill` to be controlled more granularly vs only having one global on/off switch. Autofill subfeatures handled in this PR:
- can save credentials from WebView
- can inject credentials into WebView
- can access credential management screen

### Steps to test this PR

Testing all the various scenarios is difficult because of the large number of different inputs as to whether autofill should be enabled
- including whether is an internal tester, if remote config is enabled, if autofill subfeature remote config is enabled, if the device can support it, if the user has appropriate authentication set to support it etc...
- Main test scenarios detailed here [Autofill: Remote config testing scenarios](https://app.asana.com/0/1203838575243409/1203838575243409/f)

### CODE REVIEW

**Comment by CDRussell** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #2789** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/2789" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/2789?utm_source=stack-comment).

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * Used to check the status of various Autofill features.
+ *
+ * Whether autofill features are enabled depends on a variety of inputs. This class provides a single way to query the status of all of them.
+ */
+interface AutofillCapabilityChecker {
+
+    /**
+     * Whether autofill is configured to be enabled. This is a configuration value, not a user preference.
+     */
+    suspend fun isAutofillEnabledByConfiguration(): Boolean
```
Comment (aitorvs) [inline autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt:29]: I don't think any of these APIs should be `suspend`. Suspend is not a threading mechanism and we're wrongly using to "enforce" threading.
Suspend doesn't enforce threading, and now however we're letting `suspend` to pollute the callers.

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+package com.duckduckgo.autofill.api
+
+/**
+ * Used to check the status of various Autofill features.
+ *
+ * Whether autofill features are enabled depends on a variety of inputs. This class provides a single way to query the status of all of them.
+ */
+interface AutofillCapabilityChecker {
```
Comment (aitorvs) [inline autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt:24]: I don't think any of these APIs should be `suspend`. Suspend is not a threading mechanism and we're wrongly using to "enforce" threading.
Suspend doesn't enforce threading, and now however we're letting `suspend` to pollute the callers.

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt (hunk 1 above)
Comment (aitorvs) [inline autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt:29]: This API is very odd. It's internally checking for `isInternal` and that should be left to the caller.
I would think these APIs are way simpler than they're right now. Their composability should be left to the caller imo.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    override suspend fun isAutofillEnabledByConfiguration(): Boolean = autofillGlobalCapabilityChecker.isAutofillEnabledByConfiguration()
+    override suspend fun isSecureAutofillAvailable(): Boolean = autofillGlobalCapabilityChecker.isSecureAutofillAvailable()
+    override suspend fun isAutofillEnabledByUser(): Boolean = autofillGlobalCapabilityChecker.isAutofillEnabledByUser()
+
+    override suspend fun canInjectCredentialsToWebView(): Boolean {
+        if (!isSecureAutofillAvailable()) return false
+        if (!isAutofillEnabledByConfiguration()) return false
+        if (!isAutofillEnabledByUser()) return false
+
+        if (isInternalTester()) return true
```
Comment (aitorvs) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt:48]: These can of checks are odd imo. This should be left to the caller to decide

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillGlobalCapabilityChecker.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+) : AutofillGlobalCapabilityChecker {
+
+    override suspend fun isSecureAutofillAvailable(): Boolean {
+        if (!autofillStore.autofillAvailable) return false
+        if (!deviceAuthenticator.hasValidDeviceAuthentication()) return false
+        return true
+    }
+
+    override suspend fun isAutofillEnabledByConfiguration(): Boolean {
+        return isInternalTester() || isGlobalFeatureEnabled()
```
Comment (aitorvs) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillGlobalCapabilityChecker.kt:49]: Ditto, I would remove `isInternal` checks. Should be the caller deciding that

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt (hunk 1 above)
Comment (CDRussell) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt:29]: Summarising offline chat here, in reply to all the existing comments.

the TL;DR of that class it that whether autofill can be enabled varies depending on a whole bunch of things, such as:

- internal tester
- global feature flag enabled
- subfeature flag enabled
- device capabilities
- if user has device auth set
- which autofill subfeature you're querying
- which part of the codebase you're calling from
- etc...

and it's not scalable to have each caller know about and query the right things. so the purpose of this class is to centralise and contain the complexity so that other callers don't need to consider all of the many nuanced cases.

For `isInternal` check, it's one of the many factors that determine autofill availability. if it lived outside the class (which is possible) you have to proliferate that `isInternal` check across the codebase. i think this is worse, because now every caller has to be aware of internal testing and check its state, in addition to checking other states, and combining the results. it's also optimising for deletion: when we remove the internal testing check it's only one place it is removed; the rest of the codebase is insulated from its existence or later removal.

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt (hunk 2 above)
Comment (CDRussell) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/AutofillCapabilityChecker.kt:24]: As discussed offline, there can be a larger conversation about APIs and `suspend`, but since `autofill` is already built around async, the current project is not the place to resolve that discussion or bring in wider changes to it.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt (hunk 3 above)
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt:48]: See https://github.com/duckduckgo/Android/pull/2789#discussion_r1091912371

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillGlobalCapabilityChecker.kt (hunk 4 above)
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillGlobalCapabilityChecker.kt:49]: See https://github.com/duckduckgo/Android/pull/2789#discussion_r1091912371

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/AutofillFeatureImpl.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    boundType = AutofillFeatureToggle::class,
+)
+@SingleInstanceIn(AppScope::class)
+class AutofillFeatureImpl @Inject constructor(
+    private val appBuildConfig: AppBuildConfig,
+    private val repository: AutofillFeatureToggleRepository,
+) : AutofillFeatureToggle {
+
+    override fun isFeatureEnabled(
+        featureName: AutofillSubfeatureName,
```
Comment (marcosholgado) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/AutofillFeatureImpl.kt:40]: Should be a String as discussed

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            autofillFeatureRepository.updateAllExceptions(exceptions)
+
+            val isEnabled = feature.state == "enabled"
+            autofillFeatureToggleRepository.insert(AutofillFeatureToggles(element, isEnabled, feature.minSupportedVersion))
+
+            feature.settings?.forEach { setting ->
+                Timber.v("Processing autofill setting: %s", setting.key)
+
+                when (setting.key) {
+                    "features" -> {
```
Comment (marcosholgado) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt:62]: As discussed I'll push my changes for this file.

**Review by marcosholgado — COMMENTED** [review]
@CDRussell we've talked about most of these comments already but here they are :)

Once we made the changes as discussed I'll re-review and approve.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+fun getAutofillSubfeatureElement(featureName: String): AutofillSubfeatureName? {
+    return AutofillSubfeatureName.values().find { it.value == featureName }
+}
+
+data class AutofillFeatureJson(
+    val state: String?,
+    val minSupportedVersion: Int?,
+    val exceptions: List<AutofillExceptionEntity>,
+    val settings: Map<String, JSONObject?>?,
```
[the remark this answers is not in the document]
Comment (marcosholgado) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt]: We discussed this so closing

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    private val autofillSubfeatureJsonParser: AutofillSubfeatureJsonParser,
+    private val moshi: Moshi,
+) : PrivacyFeaturePlugin {
+    override val featureName: String = AutofillFeatureName.Autofill.value
+
+    override fun store(
+        featureName: String,
+        jsonString: String,
+    ): Boolean {
+        val element = getAutofillElement(featureName) ?: return false
```
[the remark this answers is not in the document]
Comment (marcosholgado) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt]: Discussed so resolving

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    private val internalTestUserChecker: InternalTestUserChecker,
+    private val autofillGlobalCapabilityChecker: AutofillGlobalCapabilityChecker,
+) : AutofillCapabilityChecker {
+
+    override suspend fun isAutofillEnabledByConfiguration(): Boolean = autofillGlobalCapabilityChecker.isAutofillEnabledByConfiguration()
+    override suspend fun isSecureAutofillAvailable(): Boolean = autofillGlobalCapabilityChecker.isSecureAutofillAvailable()
+    override suspend fun isAutofillEnabledByUser(): Boolean = autofillGlobalCapabilityChecker.isAutofillEnabledByUser()
+
+    override suspend fun canInjectCredentialsToWebView(): Boolean {
+        if (!isSecureAutofillAvailable()) return false
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt:44]: No is correct 👍

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/configuration/AutofillRuntimeConfigProvider.kt [hunk 10]
```diff
@@ -35,9 +35,9 @@ interface AutofillRuntimeConfigProvider {
 @ContributesBinding(AppScope::class)
 class RealAutofillRuntimeConfigProvider @Inject constructor(
     private val emailManager: EmailManager,
-    private val deviceAuthenticator: DeviceAuthenticator,
     private val autofillStore: AutofillStore,
     private val runtimeConfigurationWriter: RuntimeConfigurationWriter,
+    private val autofillCapabilityChecker: AutofillCapabilityChecker,
 ) : AutofillRuntimeConfigProvider {
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/configuration/AutofillRuntimeConfigProvider.kt:41]: Initial reaction is that we already cache some of it in memory (like the initial script load) and that the hardest part to meaningfully optimise further is around being able to say "if something has changed" without having to check it all, which I don't think is trivial.

Still, it's a good question, and there might be more we can do here, so capturing as: [Android Autofill: Cache more of the autofill runtime config](https://app.asana.com/0/1203115518054842/1203884488082756/f)

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/AutofillFeatureImpl.kt (hunk 5 above)
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/AutofillFeatureImpl.kt:40]: Context (for when I later forget why we'd want to use a String instead of a stronger type...): 
- the stronger typing would be preferred were it not for us planning to codegen a bunch of this
- making it a String here is a step towards making it consistent and easier to codegen later

File: app/src/main/java/com/duckduckgo/app/settings/SettingsViewModel.kt [hunk 11]
```diff
@@ -164,7 +164,7 @@ class SettingsViewModel @Inject constructor(
                     appTrackingProtectionOnboardingShown = vpnStore.didShowOnboarding(),
                     appTrackingProtectionEnabled = vpnFeaturesRegistry.isFeatureRegistered(AppTpVpnFeature.APPTP_VPN),
                     emailAddress = emailManager.getEmailAddress(),
-                    showAutofill = autofillStore.autofillAvailable,
+                    showAutofill = autofillCapabilityChecker.canAccessCredentialManagementScreen(),
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/settings/SettingsViewModel.kt:167]: is unclear to me that the suggested alternative is better: the management part of credential management seems like a good inclusion than a more general "can access credential screen"

will leave as is since is suggested as a nit only

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/feature/AutofillFeatures.kt [hunk 12]
```diff
@@ -14,9 +14,18 @@
...
 
-package com.duckduckgo.autofill.api
+package com.duckduckgo.autofill.api.feature
 
-/** List of [AutofillFeatureName] that belong to the Autofill feature */
 enum class AutofillFeatureName(val value: String) {
     Autofill("autofill"),
 }
+
+enum class AutofillSubfeatureName(override val value: String) : AutofillSubfeature {
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/feature/AutofillFeatures.kt:23]: Not used publicly but they are used from both `autofill-impl` and `autofill-store` modules, so i think leaving it defined in `autofill-api` is ok here?

[the remark this answers is not in the document]
Comment (marcosholgado) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/feature/AutofillFeatures.kt:23]: Public means out of autofill so if there are no usages out of autofill I would move that. Worse case scenario at some point we need this public and we copy back to the api but since your public api is the capabilities checker that seems unlikely.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        if (!isAutofillEnabledByConfiguration()) return false
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt]: As discussed offline, I feel strongly that the capability checker should always return the right answer and that shouldn't depend on something having been done elsewhere. e.g., i think it's bad to base the logic here on an assumption that something will always happen elsewhere in the codebase. and would rather set this up to always give a self-contained, correct answer.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        if (!isAutofillEnabledByUser()) return false
+
+        if (isInternalTester()) return true
+
+        return autofillFeatureToggle.isFeatureEnabled(InjectCredentials, defaultValue = false)
+    }
+
+    override suspend fun canSaveCredentialsFromWebView(): Boolean {
+        if (!isSecureAutofillAvailable()) return false
+        if (!isAutofillEnabledByConfiguration()) return false
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt]: Same answer: https://github.com/duckduckgo/Android/pull/2789#discussion_r1094733540

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/configuration/AutofillRuntimeConfigProvider.kt [hunk 15]
```diff
@@ -87,5 +84,9 @@ class RealAutofillRuntimeConfigProvider @Inject constructor(
         }
     }
 
+    private suspend fun determineIfAutofillEnabled(): Boolean {
+        return autofillCapabilityChecker.isAutofillEnabledByConfiguration() && autofillCapabilityChecker.isAutofillEnabledByUser()
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/configuration/AutofillRuntimeConfigProvider.kt]: >I don't think you need to check if the config is enabled because that's already done down the line and if is not this will never be called.

Answered here: https://github.com/duckduckgo/Android/pull/2789#discussion_r1094733540



>Also autofillCapabilityChecker.isAutofillEnabledByConfiguration() calls autofillGlobalCapabilityChecker.isAutofillEnabledByConfiguration() so I would just use the latter and make the former non-public

Other parts of the codebase should not have to interrogate `autofillGlobalCapabilityChecker` (or `internalTestChecker`, or the likes...) directly in regards to autofill capabilities; they should use the one API for this, even if that is internally only deferring to something else

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+fun getAutofillElement(featureName: String): AutofillFeatureName? {
+    return AutofillFeatureName.values().find { it.value == featureName }
+}
+
+fun getAutofillSubfeatureElement(featureName: String): AutofillSubfeatureName? {
+    return AutofillSubfeatureName.values().find { it.value == featureName }
+}
+
+data class AutofillFeatureJson(
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/feature/plugin/AutofillFeaturePlugin.kt]: Agreed; this was a hangover because you and I were both making changes to the same files at the same when this work was starting.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillJavascriptInterface.kt [hunk 17]
```diff
@@ -89,6 +91,11 @@ class AutofillStoredBackJavascriptInterface @Inject constructor(
     override fun getAutofillData(requestString: String) {
         Timber.v("BrowserAutofill: getAutofillData called:\n%s", requestString)
         getAutofillDataJob += coroutineScope.launch(dispatcherProvider.default()) {
+            if (!autofillCapabilityChecker.canInjectCredentialsToWebView()) {
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillJavascriptInterface.kt]: Potentially unnecessary, if our JS is doing its job correctly. But were a regression to appear in the JS or this to be called unexpectedly i'd want our side to guard against it

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillJavascriptInterface.kt [hunk 18]
```diff
@@ -146,6 +153,11 @@ class AutofillStoredBackJavascriptInterface @Inject constructor(
         Timber.i("storeFormData called, credentials provided to be persisted")
 
         storeFormDataJob += coroutineScope.launch(dispatcherProvider.default()) {
+            if (!autofillCapabilityChecker.canSaveCredentialsFromWebView()) {
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillJavascriptInterface.kt]: https://github.com/duckduckgo/Android/pull/2789#discussion_r1094741567

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/feature/AutofillFeatures.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (marcosholgado) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/feature/AutofillFeatures.kt:23]: Discussed, we'll leave in api because it's required in the store.

[the remark this answers is not in the document]
Comment (CDRussell) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/feature/AutofillFeatures.kt:23]: Decision from conversation elsewhere: leaving it as is since it is used by `impl` and `store` (Could make the store use strings instead of the stronger typing but not worth doing)

**Review by marcosholgado — APPROVED** [review]
LGTM!

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+     * Because the credential management screen handles the states where the user has toggled autofill off, or the device can't support it,
+     * this feature is not dependent those checks.
+     *
+     * We purposely don't couple this check against [isSecureAutofillAvailable] or [isAutofillEnabledByUser].
+     */
+    override suspend fun canAccessCredentialManagementScreen(): Boolean {
+        if (isInternalTester()) return true
+        if (!isGlobalFeatureEnabled()) return false
+        return isSubfeatureEnabled(AccessCredentialManagement, defaultValue = false)
+    }
```
Comment (aitorvs) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/AutofillCapabilityCheckerImpl.kt:73]: This is merged, but just FYI, this is unnecessarily changing thread 3 times (main -> io -> main -> io -> main -> io) when called from `onShowUserCredentialsSaved`

I have measured this, the perf impact is very noticeable.
